core/src/main/scala/BinaryTree.scala

```diff
@@ -0,0 +1,248 @@
+      case BTNil()              =>  x
+      case Branch(a, l, _, _, _)      =>  loop(l, a)
+    }
+
+    this match {
+      case BTNil()                    => None()
+      case Branch(a, l, _, _, _)      =>  loop(l, a)
+    }
+  }
+
+  def max(): Option[A] = {
+    @tailrec def loop(sub: BinaryTree[A], x: Option[A]): Option[A] = sub match {
+      case BTNil()                    =>  x
+      case Branch(a, _, r, _, _)      =>  loop(r, a)
+    }
+
+    this match {
+      case BTNil()                    =>  None()
+      case Branch(a, _, r, _, _)      =>  loop(r, a)
+    }
+  }
+
+  def contains(x: A)(implicit order: Order[A]): Boolean = this match {
+    case BTNil()                      =>  false
+
+    case Branch(Some(a), l, r, _, _)  => order.compare(x, a) match {
+      case LT                             =>  l.contains(x)
+      case EQ                             =>  true
+      case GT                             =>  r.contains(x)
+    }
+  }
+
+  def add(x: A)(implicit order: Order[A]): Branch[A] = {
+    def insert(x: A, order: Order[A])  = this match {
+      case BTNil()                    =>  Branch(Some(x), BinaryTree.empty, BinaryTree.empty)
+      case Branch(Some(a), l, r, _, _)=>  order.compare(x, a) match {
+        case LT                     =>  Branch(Some(a), l.add(x)(order), r)
+        case EQ                     =>  this.asInstanceOf[Branch[A]]
+        case GT                     =>  Branch(Some(a), l, r.add(x)(order))
+      }
+    }
+
+    balance(insert(x, order)).asInstanceOf[Branch[A]]
+  }
+
+  def +(x: A)(implicit order: Order[A]): BinaryTree[A] = add(x)
+
+  def remove(x: A)(implicit order: Order[A]): BinaryTree[A] = {
+    def del(x:A)(implicit order: Order[A]): BinaryTree[A] = this match {
+      case BTNil() => BinaryTree.empty
+      case Branch(Some(a), l, r, _, _) => order.compare(x, a) match {
+        case LT => Branch(Some(a), l.remove(x), r)
+        case GT => Branch(Some(a), l, r.remove(x))
+        case EQ => (l, r) match {
+          case (BTNil(), BTNil()) => BinaryTree.empty
+          case (BTNil(), x) => x
+          case (x, BTNil()) => x
+          case (x, y) => {
+            val min = y.min
+
+            min match {
+              case Some(a) => Branch(min, x, y.remove(a))
+              case None() => x //<- this should never happen
+            }
+          }
+        }
+      }
+    }
+
+    balance(del(x))
+  }
+
+  def join(another: BinaryTree[A])(implicit order: Order[A]) = {
+
+    @tailrec def build(sub: BinaryTree[A], xs: List[A]): BinaryTree[A] = xs match {
+      case El()                 =>  sub
+      case Nel(h, t)            =>  build(sub + h, t)
+    }
+
+    another match {
+      case BTNil()              =>  this
+      case _                    =>  build(this, another.toLst())
+    }
+  }
+
+  def ++(another: BinaryTree[A])(implicit order: Order[A]) = join(another)
+
+  private def balance(tree: BinaryTree[A]): BinaryTree[A] = {
```

Comment:
balance is actaully undefined for BTNil, right? it really should be a Branch => Branch, right? what actually happens if you pass a BTNil in here?